glean-core/src/upload/mod.rs

```diff
@@ -94,98 +123,171 @@ impl PingUploadManager {
     fn test_doesnt_error_when_there_are_no_pending_pings() {
-        let data_dir = tempfile::tempdir().unwrap();
-        let tmpname = data_dir.path().display().to_string();
-
         // Create a new upload_manager
+        let dir = tempfile::tempdir().unwrap();
+        let tmpname = dir.path().display().to_string();
         let upload_manager = PingUploadManager::new(&tmpname);
 
-        // Try and get the next request
-        let request = upload_manager.get_next_ping();
+        // Wait for processing of pending pings directory to finish.
+        while upload_manager.get_next_ping() == PingUploadTask::Wait {
+            thread::sleep(Duration::from_millis(10));
+        }
 
+        // Try and get the next request.
         // Verify request was not returned
-        assert!(request.is_none());
+        assert_eq!(upload_manager.get_next_ping(), PingUploadTask::Done);
     }
 
     #[test]
     fn test_returns_ping_request_when_there_is_one() {
-        let data_dir = tempfile::tempdir().unwrap();
-        let tmpname = data_dir.path().display().to_string();
-
         // Create a new upload_manager
+        let dir = tempfile::tempdir().unwrap();
+        let tmpname = dir.path().display().to_string();
         let upload_manager = PingUploadManager::new(&tmpname);
 
-        // Enqueue a ping
-        upload_manager.enqueue_ping(UUID, URL, json!({}));
+        // Wait for processing of pending pings directory to finish.
+        while upload_manager.get_next_ping() == PingUploadTask::Wait {
+            thread::sleep(Duration::from_millis(10));
+        }
 
-        // Try and get the next request
-        let request = upload_manager.get_next_ping();
+        // Enqueue a ping
+        upload_manager.enqueue_ping(UUID, URL, json!("{}"));
 
+        // Try and get the next request.
         // Verify request was returned
-        assert!(request.is_some());
+        match upload_manager.get_next_ping() {
+            PingUploadTask::Upload(_) => {}
+            _ => panic!("Expected upload manager to return the next request!"),
+        }
     }
 
     #[test]
-    fn test_returns_as_many_ping_request_as_there_are() {
-        let data_dir = tempfile::tempdir().unwrap();
-        let tmpname = data_dir.path().display().to_string();
-
+    fn test_returns_as_many_ping_requests_as_there_are() {
         // Create a new upload_manager
+        let dir = tempfile::tempdir().unwrap();
+        let tmpname = dir.path().display().to_string();
         let upload_manager = PingUploadManager::new(&tmpname);
 
+        // Wait for processing of pending pings directory to finish.
+        while upload_manager.get_next_ping() == PingUploadTask::Wait {
+            thread::sleep(Duration::from_millis(10));
+        }
+
         // Enqueue a ping multiple times
         let n = 10;
         for _ in 0..n {
-            upload_manager.enqueue_ping(UUID, URL, json!({}));
+            upload_manager.enqueue_ping(UUID, URL, json!("{}"));
```

Comment:
`json!({})` and `json!("{}")` are different things.

The first one results in an empty json object (literally just `{}`).
the second one results in the string "{}".

We want the first one: an empty object.

Reply:
Also an error on rebasing, but I actually didn't know this.